packages/grafana-ui/src/types/panelOptions.tsx

```diff
@@ -0,0 +1,181 @@
+import { SelectOptionItem } from '../components/Select/Select';
+
+/**
+ * Option editors, i.e. BooleanOption have to comply to OptionInputAPI.
+ * OptionInputAPI represents UI elements that allow value modification, name it string, boolean, complex object, whatever
+ */
+
+export interface OptionInputAPI<T> {
+  value: T;
+  onChange: (value: T, event?: React.SyntheticEvent<HTMLElement>) => void;
+}
+
+export interface OptionsRowModel<TOptions> {
+  columns: number;
+  content: Array<OptionUIModel<TOptions, any> | GroupLayoutUIModel<any>>;
+}
+export interface OptionsUIModel<TOptions> {
+  rows: Array<OptionsRowModel<TOptions>>;
+}
+
+export enum OptionType {
+  Text = 'text',
+  Float = 'float',
+  Integer = 'integer',
+  Boolean = 'boolean',
+  Object = 'object',
+  Array = 'array',
+}
+
+/**
+ * Introducing this interface to make sure component for manipulating given option
+ * handles the type that's specific for that option.
+ * i.e. for maxValue: number we expect the component to accept value property of that type
+ * and onChange handler that exposes modified value of that type.
+ *
+ * onChange has to accept event property as well. Exposing event as a second argument is purposeful.
+ * Driver for that is that the primary expected output from InputAPI component is it's modified value rather than the event object.
+ * But, event is also very important in certain cases, hence we need to keep it in the API.
+ */
+export interface OptionUIComponentProps<T> extends OptionInputAPI<T> {}
+
+export interface SelectOptionUIComponentProps<T> extends OptionInputAPI<SelectOptionItem<T>> {
+  options: Array<SelectOptionItem<T>>;
+}
+
+interface AbstractOptionUIModel<ComponentProps, P> {
+  type: OptionType;
+  /**
+   * Name of the options object property
+   * i.e. for GaugeOptions type:
+   * export const defaults: GaugeOptions = {
+   *   minValue: 0,
+   *   maxValue: 100,
+   *   ...
+   * }
+   * it could be 'minValue'
+   * Question: Should we name it option instead of path?
+   */
+  path: P;
```

Comment:
Not sure we should name it `path` since I'm allowing shallow options manipulation only. I think we should strive for editors composition(via declarative API, i.e. `useCustomOption('Some custom options', 'customOptionKey`, CustomOptionEditor)` rather than handling deep paths for options. Any thoughts on that @ryantxu?

Reply:
Does `path` represent the json attribute name?  if so, maybe `property`?  and make sure it is a valid property

Reply:
Yes, it's basically json attribute name. `property` sgtm